R/nest.R

```diff
@@ -106,39 +156,52 @@ nest <- function(.data, ..., .names_sep = NULL, .key = deprecated()) {
       i = "Did you want `{(.key)} = {cols_fixed_label}`?"
     ))
 
-    return(nest(.data, !!!cols))
+    return(nest(.data, !!!cols, .by = {{ .by }}))
   }
 
   UseMethod("nest")
 }
 
 #' @export
-nest.data.frame <- function(.data, ..., .names_sep = NULL, .key = deprecated()) {
+nest.data.frame <- function(.data,
+                            ...,
+                            .by = NULL,
+                            .key = NULL,
+                            .names_sep = NULL) {
   # The data frame print handles nested data frames poorly, so we want to
   # convert data frames (but not subclasses) to tibbles
   if (identical(class(.data), "data.frame")) {
     .data <- as_tibble(.data)
   }
 
-  nest.tbl_df(.data, ..., .names_sep = .names_sep, .key = .key)
+  nest.tbl_df(
+    .data,
+    ...,
+    .by = {{ .by }},
+    .key = .key,
+    .names_sep = .names_sep
+  )
 }
 
 #' @export
-nest.tbl_df <- function(.data, ..., .names_sep = NULL, .key = deprecated()) {
-  .key <- check_key(.key)
-  if (missing(...)) {
-    cli::cli_warn(c(
-      "`...` can't be empty for ungrouped data frames.",
-      i = "Did you want `{(.key)} = everything()`?"
-    ))
-    cols <- quos(!!.key := everything())
-  } else {
-    cols <- enquos(...)
-  }
-
+nest.tbl_df <- function(.data,
+                        ...,
+                        .by = NULL,
+                        .key = NULL,
+                        .names_sep = NULL) {
   error_call <- current_env()
 
-  out <- pack(.data, !!!cols, .names_sep = .names_sep, .error_call = error_call)
+  info <- nest_info(.data, ..., .by = {{ .by }}, .key = .key)
+  cols <- info$cols
+  inner <- info$inner
+  outer <- info$outer
+
+  inner <- .data[inner]
+  inner <- pack(inner, !!!cols, .names_sep = .names_sep, .error_call = error_call)
+
+  out <- .data[outer]
+  out <- vec_cbind(out, inner, .name_repair = "check_unique", .error_call = error_call)
+  out <- reconstruct_tibble(.data, out)
```

Comment:
`vec_cbind()` with a grouped data frame will drop the `<grouped_df>` class and return a bare tibble. So we use `reconstruct_tibble()` after that to restore the grouped class if possible. There are tests for this.

Both `pack()` and `chop()` also do `reconstruct_tibble()` on the way out so we don't need it after them